packages/langium/test/languages/domain-model/lsp/goto-resolver.test.ts

```diff
@@ -0,0 +1,76 @@
+/******************************************************************************
+ * Copyright 2021 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { Diagnostic, Position } from 'vscode-languageserver';
+import { buildRealpath, createLangiumGrammarServices, LangiumDocument, LangiumDocumentConfiguration, LangiumGrammarServices, loadUriContent } from '../../../../src';
+
+let services: LangiumGrammarServices;
+let document: LangiumDocument;
+
+beforeAll(() => {
+    const realPathFile = buildRealpath('packages/langium', 'test/languages/domain-model/domain-model.langium');
+    const grammarFileContent = loadUriContent(realPathFile);
+    document = LangiumDocumentConfiguration.create(`file:${realPathFile}`, 'langium', 0, grammarFileContent);
+    expect(document).toBeDefined;
+
+    services = createLangiumGrammarServices();
+    const diagnostics: Diagnostic[] = [];
+    expect(services.documents.DocumentBuilder.build(document)).toBeTruthy;
+    expect(diagnostics.length).toEqual(0);
+});
+
+describe('GoToResolver', () => {
+    test('Find Declaration', () => {
+        const goToResolver = services.lsp.GoToResolver;
+        const textPos = {
+            textDocument: {
+                uri: document.uri
+            },
+            // `    DataType | Entity;` - Must find definiton of Entity
+            position: Position.create(15, 16)
+        };
+        const locationLink = goToResolver.goToDefinition(document, textPos);
+        expect(locationLink.length).toBe(1);
+        expect(locationLink[0].targetSelectionRange.start.line).toBe(20);
+        expect(locationLink[0].targetSelectionRange.start.character).toBe(0);
+        expect(locationLink[0].targetSelectionRange.end.line).toBe(20);
+        expect(locationLink[0].targetSelectionRange.end.character).toBe(6);
```

Comment:
I thought about a more modular approach to testing. For example, I'd like to see a testing service somewhere in the `langium` package (or a new `langium-test` package, although I'm against that), that can be used like this:

```ts
test('Find declaration', () => {
    langiumTest.expectGoToDefinition('<input-text>', position, {
        start: {
            line: 20,
            character: 0
        },
        end: {
            line: 20,
            character: 6
        }
    });
});
```

Where `expectGoToDefinition` is something like this `expectGoToDefinition(input: string, position: Position, expected: Range)`. Maybe we need to initialize it with some kind of `expect` method, so that it can be used with other test frameworks than jest.

Reply:
For that we would also need an equivalent of the `ParseHelper` from Xtext in Langium. This should now be easily possible, since the language meta data is included in the `LangiumServices`

Reply:
While this would generally improve test readability and maintenance (as everything is tested in one place), I'm mostly concerned with downstream tests. e.g. A developer trying to test their own language/framework based on Langium. Just as with the rest of the framework, we want to provide the developer with a nice and handy API to test their code.

Reply:
Ok, so something like a collection of tests (select-what-you-need, extendable) that you execute in the context of a specific language. On every call (what you outlined above) you formulate the expectations. This could even be driven completely with configuration.